Modules/Devices/src/Devices.Application/Devices/Commands/DeleteDevice/Handler.cs

```diff
@@ -12,35 +13,27 @@ namespace Backbone.Modules.Devices.Application.Devices.Commands.DeleteDevice;
 public class Handler : IRequestHandler<DeleteDeviceCommand>
 {
     private readonly ChallengeValidator _challengeValidator;
-    private readonly IDevicesDbContext _dbContext;
     private readonly ILogger<Handler> _logger;
     private readonly IUserContext _userContext;
+    private readonly IDevicesRepository _devicesRepository;
 
-    public Handler(IDevicesDbContext dbContext, IUserContext userContext, ChallengeValidator challengeValidator, ILogger<Handler> logger)
+    public Handler(IDevicesRepository devicesRepository, IUserContext userContext, ChallengeValidator challengeValidator, ILogger<Handler> logger)
     {
-        _dbContext = dbContext;
+        _devicesRepository = devicesRepository;
         _userContext = userContext;
         _challengeValidator = challengeValidator;
         _logger = logger;
     }
 
     public async Task Handle(DeleteDeviceCommand request, CancellationToken cancellationToken)
     {
-        var device = await _dbContext.Set<Device>()
-            .OfIdentity(_userContext.GetAddress())
-            .NotDeleted()
-            .Include(d => d.Identity)
-            .FirstWithId(request.DeviceId, cancellationToken);
+        var device = await _devicesRepository.GetDeviceByIdentityAndId(_userContext.GetAddress(), request.DeviceId, cancellationToken);
 
         await _challengeValidator.Validate(request.SignedChallenge, PublicKey.FromBytes(device.Identity.PublicKey));
 
         _logger.LogTrace("Challenge successfully validated.");
 
-        device.MarkAsDeleted(request.DeletionCertificate, _userContext.GetDeviceId());
-
-        _dbContext.Set<Device>().Update(device);
-
-        await _dbContext.SaveChangesAsync(cancellationToken);
+        await _devicesRepository.MarkAsDeleted(device.Id, request.DeletionCertificate, _userContext.GetDeviceId(), cancellationToken);
```

Comment:
Try to adhere to the following order when it comes to placing the application logic:
1. Domain
2. Application
3. Infrastructure

In other words: only place application logic into the Infrastructure project if it is absolutely necessary (the only reason for this is usually performance optimizations).

So in this example:

```suggestion
        device.MarkAsDeleted(...); // this is application logic and should therefore be located in the Application project.
        await _devicesRepository.Update(device);
```